sources/packages/backend/libs/utilities/src/application-data-comparison/application-data-comparison.ts

```diff
@@ -0,0 +1,168 @@
+    return;
+  }
+  // Property has a null or undefined value but had a value before or vice versa.
+  if (
+    (currentData === null && previousData !== null) ||
+    (currentData !== null && previousData === null) ||
+    (currentData === undefined && previousData !== undefined) ||
+    (currentData !== undefined && previousData === undefined)
+  ) {
+    const newValueChange = new ApplicationDataChange({
+      key: options?.propertyKey,
+      newValue: currentData,
+      oldValue: previousData,
+      index: options?.index,
+    });
+    parentChange.changes.push(newValueChange);
+    return;
+  }
+  // Check array items.
+  // previousData is checked also to enforce the type as an array. In case
+  // it is null or undefined it will be handled at the previous scenario.
+  if (Array.isArray(currentData) && Array.isArray(previousData)) {
+    checkArrayChanges(currentData, previousData, parentChange, options);
+    return;
+  }
+  // Check object properties.
+  if (typeof currentData === "object") {
+    checkObjectChanges(currentData, previousData, parentChange, options);
+    return;
+  }
+  // Property was changed and it is a leaf property.
+  const newChange = new ApplicationDataChange({
+    key: options?.propertyKey,
+    newValue: currentData,
+    oldValue: previousData,
+    index: options?.index,
+  });
+  parentChange.changes.push(newChange);
+}
+
+/**
+ * Checks if an array has changes comparing the currentData and previousData arrays.
+ * @param currentData current data array.
+ * @param previousData previous data array.
+ * @param parentChange parent change that will have the new {@link ApplicationDataChange} added.
+ * @param options options to support the new change creation.
+ * - `propertyKey`: key of the property that changed.
+ * - `index`: index of the array item that changed.
+ */
+function checkArrayChanges(
+  currentData: unknown[],
+  previousData: unknown[],
+  parentChange: ApplicationDataChange,
+  options?: { propertyKey?: string; index?: number },
+): void {
+  let arrayItemChange: ApplicationDataChange;
+  if (!isEqual(currentData, previousData)) {
+    arrayItemChange = new ApplicationDataChange({ key: options?.propertyKey });
+    parentChange.changes.push(arrayItemChange);
+    if (currentData.length < previousData.length) {
+      arrayItemChange.changeType = ChangeTypes.ItemsRemoved;
+    } else if (currentData.length > previousData.length) {
+      arrayItemChange.changeType = ChangeTypes.ItemsAppended;
+    }
+  }
+  // Property is an array that should have its items checked.
+  for (let index = 0; index < currentData.length; index++) {
+    const currentDataArrayItem = currentData[index];
+    const previousDataArrayItem = previousData[index];
+    if (!isEqual(currentDataArrayItem, previousDataArrayItem)) {
+      compareApplicationDataRecursive(
+        currentDataArrayItem,
+        previousDataArrayItem,
+        arrayItemChange,
+        { index },
+      );
+    }
+  }
```

Comment:
Isn't the for suppose to come inside the if consition?

Reply:
Yes, you were right. Removed as per our work session. Nice catch 😉